File: test/ModularPipelines.UnitTests/DynamicDependencyDeclarationTests.cs
using ModularPipelines.Attributes;
using ModularPipelines.Context;
using ModularPipelines.Enums;
using ModularPipelines.Exceptions;
using ModularPipelines.Models;
using ModularPipelines.Modules;
using ModularPipelines.TestHelpers;
using Status = ModularPipelines.Enums.Status;

namespace ModularPipelines.UnitTests;

/// <summary>
/// Tests for the dynamic dependency declaration feature introduced in Issue #1870.
/// </summary>
public class DynamicDependencyDeclarationTests : TestBase
{
    #region Helper Modules

    /// <summary>
    /// A basic module with no dependencies.
    /// </summary>
    private class BaseModule : SimpleTestModule<string>
    {
        protected override string Result => "base";
    }

    /// <summary>
    /// A module that others can optionally depend on.
    /// </summary>
    private class OptionalDependencyModule : SimpleTestModule<string>
    {
        protected override string Result => "optional";
    }

    /// <summary>
    /// A module for testing lazy dependencies.
    /// </summary>
    private class LazyModule : SimpleTestModule<string>
    {
        protected override string Result => "lazy";
    }

    /// <summary>
    /// A module for testing conditional dependencies.
    /// </summary>
    private class ConditionalModule : SimpleTestModule<string>
    {
        protected override string Result => "conditional";
    }

    /// <summary>
    /// A module that declares a required dependency programmatically.
    /// </summary>
    private class ModuleWithProgrammaticDependency : Module<string>
    {
        protected override void DeclareDependencies(IDependencyDeclaration deps)
        {
            deps.DependsOn<BaseModule>();
        }

        public override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return "programmatic";
        }
    }

    /// <summary>
    /// A module that declares an optional dependency programmatically.
    /// </summary>
    private class ModuleWithOptionalDependency : Module<string>
    {
        protected override void DeclareDependencies(IDependencyDeclaration deps)
        {
            deps.DependsOnOptional<OptionalDependencyModule>();
        }

        public override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return "optional-dep";
        }
    }

    /// <summary>
    /// A module that declares a conditional dependency that is active.
    /// </summary>
    private class ModuleWithActiveConditionalDependency : Module<string>
    {
        protected override void DeclareDependencies(IDependencyDeclaration deps)
        {
            deps.DependsOnIf<ConditionalModule>(true);
        }

        public override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return "conditional-active";
        }
    }

    /// <summary>
    /// A module that declares a conditional dependency that is inactive.
    /// </summary>
    private class ModuleWithInactiveConditionalDependency : Module<string>
    {
        protected override void DeclareDependencies(IDependencyDeclaration deps)
        {
            deps.DependsOnIf<ConditionalModule>(false);
        }

        public override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return "conditional-inactive";
        }
    }

    /// <summary>
    /// A module that declares a conditional dependency using a predicate.
    /// </summary>
    private class ModuleWithPredicateConditionalDependency : Module<string>
    {
        public static bool ShouldDependOnConditional { get; set; } = true;

        protected override void DeclareDependencies(IDependencyDeclaration deps)
        {
            deps.DependsOnIf<ConditionalModule>(() => ShouldDependOnConditional);
        }

        public override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return "predicate-conditional";
        }
    }

    /// <summary>
    /// A module that declares a lazy dependency.
    /// </summary>
    private class ModuleWithLazyDependency : Module<string>
    {
        protected override void DeclareDependencies(IDependencyDeclaration deps)
        {
            deps.DependsOnLazy<LazyModule>();
        }

        public override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return "lazy-dep";
        }
    }

    /// <summary>
    /// A module that combines both attribute and programmatic dependencies.
    /// </summary>
    [ModularPipelines.Attributes.DependsOn<BaseModule>]
    private class ModuleWithBothAttributeAndProgrammaticDependencies : Module<string>
    {
        protected override void DeclareDependencies(IDependencyDeclaration deps)
        {
            deps.DependsOnOptional<OptionalDependencyModule>();
        }

        public override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return "combined";
        }
    }

    /// <summary>
    /// A module that chains multiple dependency declarations.
    /// </summary>
    private class ModuleWithChainedDependencies : Module<string>
    {
        protected override void DeclareDependencies(IDependencyDeclaration deps)
        {
            deps.DependsOn<BaseModule>()
                .DependsOnOptional<OptionalDependencyModule>()
                .DependsOnLazy<LazyModule>();
        }

        public override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return "chained";
        }
    }

    /// <summary>
    /// A module that depends on an unregistered required dependency (should fail).
    /// </summary>
    private class ModuleWithMissingRequiredDependency : Module<string>
    {
        protected override void DeclareDependencies(IDependencyDeclaration deps)
        {
            deps.DependsOn<BaseModule>(); // BaseModule not registered
        }

        public override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return "missing-dep";
        }
    }

    /// <summary>
    /// A module that uses DependsOn with Type parameter.
    /// </summary>
    private class ModuleWithTypeDependency : Module<string>
    {
        protected override void DeclareDependencies(IDependencyDeclaration deps)
        {
            deps.DependsOn(typeof(BaseModule));
        }

        public override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Yield();
            return "type-dep";
        }
    }

    #endregion

    #region Required Dependency Tests

    [Test]
    public async Task Programmatic_Required_Dependency_Works_When_Registered()
    {
        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<BaseModule>()
            .AddModule<ModuleWithProgrammaticDependency>()
            .ExecutePipelineAsync();

        await Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful);
    }

    [Test]
    public async Task Programmatic_Required_Dependency_Throws_When_Not_Registered()
    {
        await Assert.That(async () => await TestPipelineHostBuilder.Create()
                .AddModule<ModuleWithMissingRequiredDependency>()
                .ExecutePipelineAsync())
            .ThrowsException();
    }

    [Test]
    public async Task Programmatic_Type_Dependency_Works()
    {
        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<BaseModule>()
            .AddModule<ModuleWithTypeDependency>()
            .ExecutePipelineAsync();

        await Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful);
    }

    #endregion

    #region Optional Dependency Tests

    [Test]
    public async Task Optional_Dependency_Works_When_Registered()
    {
        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<OptionalDependencyModule>()
            .AddModule<ModuleWithOptionalDependency>()
            .ExecutePipelineAsync();

        await Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful);
    }

    [Test]
    public async Task Optional_Dependency_Does_Not_Fail_When_Not_Registered()
    {
        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<ModuleWithOptionalDependency>()
            .ExecutePipelineAsync();

        await Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful);
    }

    #endregion

    #region Conditional Dependency Tests

    [Test]
    public async Task Conditional_Dependency_Works_When_Condition_True_And_Registered()
    {
        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<ConditionalModule>()
            .AddModule<ModuleWithActiveConditionalDependency>()
            .ExecutePipelineAsync();

        await Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful);
    }

    [Test]
    public async Task Conditional_Dependency_Throws_When_Condition_True_And_Not_Registered()
    {
        await Assert.That(async () => await TestPipelineHostBuilder.Create()
                .AddModule<ModuleWithActiveConditionalDependency>()
                .ExecutePipelineAsync())
            .ThrowsException();
    }

    [Test]
    public async Task Conditional_Dependency_Not_Added_When_Condition_False()
    {
        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<ModuleWithInactiveConditionalDependency>()
            .ExecutePipelineAsync();

        await Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful);
    }

    [Test]
    [TUnit.Core.NotInParallel(Order = 1)]
    public async Task Conditional_Predicate_Dependency_Works_When_Predicate_Returns_True()
    {
        ModuleWithPredicateConditionalDependency.ShouldDependOnConditional = true;

        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<ConditionalModule>()
            .AddModule<ModuleWithPredicateConditionalDependency>()
            .ExecutePipelineAsync();

        await Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful);
    }

    [Test]
    [TUnit.Core.NotInParallel(Order = 2)]
    public async Task Conditional_Predicate_Dependency_Not_Added_When_Predicate_Returns_False()
    {
        ModuleWithPredicateConditionalDependency.ShouldDependOnConditional = false;

        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<ModuleWithPredicateConditionalDependency>()
            .ExecutePipelineAsync();

        await Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful);
    }

    #endregion

    #region Lazy Dependency Tests

    [Test]
    public async Task Lazy_Dependency_Does_Not_Fail_When_Not_Registered()
    {
        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<ModuleWithLazyDependency>()
            .ExecutePipelineAsync();

        await Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful);
    }

    [Test]
    public async Task Lazy_Dependency_Works_When_Registered()
    {
        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<LazyModule>()
            .AddModule<ModuleWithLazyDependency>()
            .ExecutePipelineAsync();

        await Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful);
    }

    #endregion

    #region Combined Dependency Tests

    [Test]
    public async Task Combined_Attribute_And_Programmatic_Dependencies_Work()
    {
        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<BaseModule>()
            .AddModule<OptionalDependencyModule>()
            .AddModule<ModuleWithBothAttributeAndProgrammaticDependencies>()
            .ExecutePipelineAsync();

        await Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful);
    }

    [Test]
    public async Task Combined_Dependencies_Work_With_Only_Attribute_Dependency_Registered()
    {
        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<BaseModule>()
            .AddModule<ModuleWithBothAttributeAndProgrammaticDependencies>()
            .ExecutePipelineAsync();

        await Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful);
    }

    [Test]
    public async Task Chained_Dependency_Declarations_Work()
    {
        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<BaseModule>()
            .AddModule<OptionalDependencyModule>()
            .AddModule<LazyModule>()
            .AddModule<ModuleWithChainedDependencies>()
            .ExecutePipelineAsync();

        await Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful);
    }

    [Test]
    public async Task Chained_Dependency_Declarations_Work_With_Only_Required_Registered()
    {
        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<BaseModule>()
            .AddModule<ModuleWithChainedDependencies>()
            .ExecutePipelineAsync();

        await Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful);
    }

    #endregion

    #region DependencyDeclaration Unit Tests

    [Test]
    public async Task DependencyDeclaration_DependsOn_Returns_Same_Instance_For_Chaining()
    {
        var declaration = new DependencyDeclaration();

        var result = declaration.DependsOn<BaseModule>();

        await Assert.That(result).IsSameReferenceAs(declaration);
    }

    [Test]
    public async Task DependencyDeclaration_DependsOnOptional_Returns_Same_Instance_For_Chaining()
    {
        var declaration = new DependencyDeclaration();

        var result = declaration.DependsOnOptional<BaseModule>();

        await Assert.That(result).IsSameReferenceAs(declaration);
    }

    [Test]
    public async Task DependencyDeclaration_DependsOnIf_Returns_Same_Instance_For_Chaining()
    {
        var declaration = new DependencyDeclaration();

        var result = declaration.DependsOnIf<BaseModule>(true);

        await Assert.That(result).IsSameReferenceAs(declaration);
    }

    [Test]
    public async Task DependencyDeclaration_DependsOnLazy_Returns_Same_Instance_For_Chaining()
    {
        var declaration = new DependencyDeclaration();

        var result = declaration.DependsOnLazy<BaseModule>();

        await Assert.That(result).IsSameReferenceAs(declaration);
    }

    [Test]
    public async Task DependencyDeclaration_Throws_For_Non_Module_Type()
    {
        var declaration = new DependencyDeclaration();

        await Assert.That(() => declaration.DependsOn(typeof(string)))
            .ThrowsException()
            .And.HasMessageContaining("is not a Module");
    }

    [Test]
    public async Task DependencyDeclaration_Required_Has_Correct_DependencyType()
    {
        var declaration = new DependencyDeclaration();
        declaration.DependsOn<BaseModule>();

        var deps = declaration.Dependencies;

        await Assert.That(deps).HasCount().EqualTo(1);
        await Assert.That(deps[0].Kind).IsEqualTo(DependencyType.Required);
        await Assert.That(deps[0].IgnoreIfNotRegistered).IsEqualTo(false);
    }

    [Test]
    public async Task DependencyDeclaration_Optional_Has_Correct_DependencyType()
    {
        var declaration = new DependencyDeclaration();
        declaration.DependsOnOptional<BaseModule>();

        var deps = declaration.Dependencies;

        await Assert.That(deps).HasCount().EqualTo(1);
        await Assert.That(deps[0].Kind).IsEqualTo(DependencyType.Optional);
        await Assert.That(deps[0].IgnoreIfNotRegistered).IsEqualTo(true);
    }

    [Test]
    public async Task DependencyDeclaration_Lazy_Has_Correct_DependencyType()
    {
        var declaration = new DependencyDeclaration();
        declaration.DependsOnLazy<BaseModule>();

        var deps = declaration.Dependencies;

        await Assert.That(deps).HasCount().EqualTo(1);
        await Assert.That(deps[0].Kind).IsEqualTo(DependencyType.Lazy);
        await Assert.That(deps[0].IgnoreIfNotRegistered).IsEqualTo(true);
    }

    [Test]
    public async Task DependencyDeclaration_Conditional_Has_Correct_DependencyType()
    {
        var declaration = new DependencyDeclaration();
        declaration.DependsOnIf<BaseModule>(true);

        var deps = declaration.Dependencies;

        await Assert.That(deps).HasCount().EqualTo(1);
        await Assert.That(deps[0].Kind).IsEqualTo(DependencyType.Conditional);
        await Assert.That(deps[0].IgnoreIfNotRegistered).IsEqualTo(false);
    }

    [Test]
    public async Task DependencyDeclaration_Conditional_False_Does_Not_Add_Dependency()
    {
        var declaration = new DependencyDeclaration();
        declaration.DependsOnIf<BaseModule>(false);

        var deps = declaration.Dependencies;

        await Assert.That(deps).HasCount().EqualTo(0);
    }

    #endregion
}

File: test/ModularPipelines.UnitTests.FSharp/Dependencies/DynamicDependencyDeclarationTests.fs
namespace ModularPipelines.UnitTests.FSharp.Dependencies

open System
open System.Threading
open System.Threading.Tasks
open ModularPipelines.Attributes
open ModularPipelines.Context
open ModularPipelines.Enums
open ModularPipelines.Exceptions
open ModularPipelines.Extensions
open ModularPipelines.Models
open ModularPipelines.Modules
open ModularPipelines.TestHelpers
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

module DynamicDependencyDeclarationTestsData =
    type BaseModule() =
        inherit SimpleTestModule<string>()
        override _.Result = "base"

    type OptionalDependencyModule() =
        inherit SimpleTestModule<string>()
        override _.Result = "optional"

    type LazyModule() =
        inherit SimpleTestModule<string>()
        override _.Result = "lazy"

    type ConditionalModule() =
        inherit SimpleTestModule<string>()
        override _.Result = "conditional"

    type ModuleWithProgrammaticDependency() =
        inherit Module<string>()

        override _.DeclareDependencies(deps: IDependencyDeclaration) =
            deps.DependsOn<BaseModule>() |> ignore

        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            task {
                do! Task.Yield()
                return "programmatic"
            }

    type ModuleWithOptionalDependency() =
        inherit Module<string>()

        override _.DeclareDependencies(deps: IDependencyDeclaration) =
            deps.DependsOnOptional<OptionalDependencyModule>() |> ignore

        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            task {
                do! Task.Yield()
                return "optional-dep"
            }

    type ModuleWithActiveConditionalDependency() =
        inherit Module<string>()

        override _.DeclareDependencies(deps: IDependencyDeclaration) =
            deps.DependsOnIf<ConditionalModule>(true) |> ignore

        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            task {
                do! Task.Yield()
                return "conditional-active"
            }

    type ModuleWithInactiveConditionalDependency() =
        inherit Module<string>()

        override _.DeclareDependencies(deps: IDependencyDeclaration) =
            deps.DependsOnIf<ConditionalModule>(false) |> ignore

        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            task {
                do! Task.Yield()
                return "conditional-inactive"
            }

    type ModuleWithPredicateConditionalDependency() =
        inherit Module<string>()

        static member val ShouldDependOnConditional = true with get, set

        override _.DeclareDependencies(deps: IDependencyDeclaration) =
            deps.DependsOnIf<ConditionalModule>(fun () -> ModuleWithPredicateConditionalDependency.ShouldDependOnConditional)
            |> ignore

        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            task {
                do! Task.Yield()
                return "predicate-conditional"
            }

    type ModuleWithLazyDependency() =
        inherit Module<string>()

        override _.DeclareDependencies(deps: IDependencyDeclaration) =
            deps.DependsOnLazy<LazyModule>() |> ignore

        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            task {
                do! Task.Yield()
                return "lazy-dep"
            }

    [<DependsOn(typeof<BaseModule>)>]
    type ModuleWithBothAttributeAndProgrammaticDependencies() =
        inherit Module<string>()

        override _.DeclareDependencies(deps: IDependencyDeclaration) =
            deps.DependsOnOptional<OptionalDependencyModule>() |> ignore

        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            task {
                do! Task.Yield()
                return "combined"
            }

    type ModuleWithChainedDependencies() =
        inherit Module<string>()

        override _.DeclareDependencies(deps: IDependencyDeclaration) =
            deps
                .DependsOn<BaseModule>()
                .DependsOnOptional<OptionalDependencyModule>()
                .DependsOnLazy<LazyModule>()
            |> ignore

        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            task {
                do! Task.Yield()
                return "chained"
            }

    type ModuleWithMissingRequiredDependency() =
        inherit Module<string>()

        override _.DeclareDependencies(deps: IDependencyDeclaration) =
            deps.DependsOn<BaseModule>() |> ignore

        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            task {
                do! Task.Yield()
                return "missing-dep"
            }

    type ModuleWithTypeDependency() =
        inherit Module<string>()

        override _.DeclareDependencies(deps: IDependencyDeclaration) =
            deps.DependsOn(typeof<BaseModule>) |> ignore

        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            task {
                do! Task.Yield()
                return "type-dep"
            }

type DynamicDependencyDeclarationTests() =
    inherit TestBase()

    [<Test>]
    member _.Programmatic_Required_Dependency_Works_When_Registered() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<DynamicDependencyDeclarationTestsData.BaseModule>()
                .AddModule<DynamicDependencyDeclarationTestsData.ModuleWithProgrammaticDependency>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.Programmatic_Required_Dependency_Throws_When_Not_Registered() = async {
        let mutable threw = false

        try
            let! _ =
                TestPipelineHostBuilder.Create()
                    .AddModule<DynamicDependencyDeclarationTestsData.ModuleWithMissingRequiredDependency>()
                    .ExecutePipelineAsync()
                |> Async.AwaitTask
            ()
        with _ ->
            threw <- true

        do! check(Assert.That(threw).IsTrue())
    }

    [<Test>]
    member _.Programmatic_Type_Dependency_Works() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<DynamicDependencyDeclarationTestsData.BaseModule>()
                .AddModule<DynamicDependencyDeclarationTestsData.ModuleWithTypeDependency>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.Optional_Dependency_Works_When_Registered() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<DynamicDependencyDeclarationTestsData.OptionalDependencyModule>()
                .AddModule<DynamicDependencyDeclarationTestsData.ModuleWithOptionalDependency>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.Optional_Dependency_Does_Not_Fail_When_Not_Registered() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<DynamicDependencyDeclarationTestsData.ModuleWithOptionalDependency>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.Conditional_Dependency_Works_When_Condition_True_And_Registered() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<DynamicDependencyDeclarationTestsData.ConditionalModule>()
                .AddModule<DynamicDependencyDeclarationTestsData.ModuleWithActiveConditionalDependency>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.Conditional_Dependency_Throws_When_Condition_True_And_Not_Registered() = async {
        let mutable threw = false

        try
            let! _ =
                TestPipelineHostBuilder.Create()
                    .AddModule<DynamicDependencyDeclarationTestsData.ModuleWithActiveConditionalDependency>()
                    .ExecutePipelineAsync()
                |> Async.AwaitTask
            ()
        with _ ->
            threw <- true

        do! check(Assert.That(threw).IsTrue())
    }

    [<Test>]
    member _.Conditional_Dependency_Not_Added_When_Condition_False() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<DynamicDependencyDeclarationTestsData.ModuleWithInactiveConditionalDependency>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    [<NotInParallel(Order = 1)>]
    member _.Conditional_Predicate_Dependency_Works_When_Predicate_Returns_True() = async {
        DynamicDependencyDeclarationTestsData.ModuleWithPredicateConditionalDependency.ShouldDependOnConditional <- true

        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<DynamicDependencyDeclarationTestsData.ConditionalModule>()
                .AddModule<DynamicDependencyDeclarationTestsData.ModuleWithPredicateConditionalDependency>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    [<NotInParallel(Order = 2)>]
    member _.Conditional_Predicate_Dependency_Not_Added_When_Predicate_Returns_False() = async {
        DynamicDependencyDeclarationTestsData.ModuleWithPredicateConditionalDependency.ShouldDependOnConditional <- false

        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<DynamicDependencyDeclarationTestsData.ModuleWithPredicateConditionalDependency>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.Lazy_Dependency_Does_Not_Fail_When_Not_Registered() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<DynamicDependencyDeclarationTestsData.ModuleWithLazyDependency>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.Lazy_Dependency_Works_When_Registered() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<DynamicDependencyDeclarationTestsData.LazyModule>()
                .AddModule<DynamicDependencyDeclarationTestsData.ModuleWithLazyDependency>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.Combined_Attribute_And_Programmatic_Dependencies_Work() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<DynamicDependencyDeclarationTestsData.BaseModule>()
                .AddModule<DynamicDependencyDeclarationTestsData.OptionalDependencyModule>()
                .AddModule<DynamicDependencyDeclarationTestsData.ModuleWithBothAttributeAndProgrammaticDependencies>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.Combined_Dependencies_Work_With_Only_Attribute_Dependency_Registered() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<DynamicDependencyDeclarationTestsData.BaseModule>()
                .AddModule<DynamicDependencyDeclarationTestsData.ModuleWithBothAttributeAndProgrammaticDependencies>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.Chained_Dependency_Declarations_Work() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<DynamicDependencyDeclarationTestsData.BaseModule>()
                .AddModule<DynamicDependencyDeclarationTestsData.OptionalDependencyModule>()
                .AddModule<DynamicDependencyDeclarationTestsData.LazyModule>()
                .AddModule<DynamicDependencyDeclarationTestsData.ModuleWithChainedDependencies>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.Chained_Dependency_Declarations_Work_With_Only_Required_Registered() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create()
                .AddModule<DynamicDependencyDeclarationTestsData.BaseModule>()
                .AddModule<DynamicDependencyDeclarationTestsData.ModuleWithChainedDependencies>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(pipelineSummary.Status).IsEqualTo(Status.Successful))
    }

    [<Test>]
    member _.DependencyDeclaration_DependsOn_Returns_Same_Instance_For_Chaining() = async {
        let declaration = DependencyDeclaration()
        let result = declaration.DependsOn<DynamicDependencyDeclarationTestsData.BaseModule>()
        do! check(Assert.That(Object.ReferenceEquals(result, declaration)).IsTrue())
    }

    [<Test>]
    member _.DependencyDeclaration_DependsOnOptional_Returns_Same_Instance_For_Chaining() = async {
        let declaration = DependencyDeclaration()
        let result = declaration.DependsOnOptional<DynamicDependencyDeclarationTestsData.BaseModule>()
        do! check(Assert.That(Object.ReferenceEquals(result, declaration)).IsTrue())
    }

    [<Test>]
    member _.DependencyDeclaration_DependsOnIf_Returns_Same_Instance_For_Chaining() = async {
        let declaration = DependencyDeclaration()
        let result = declaration.DependsOnIf<DynamicDependencyDeclarationTestsData.BaseModule>(true)
        do! check(Assert.That(Object.ReferenceEquals(result, declaration)).IsTrue())
    }

    [<Test>]
    member _.DependencyDeclaration_DependsOnLazy_Returns_Same_Instance_For_Chaining() = async {
        let declaration = DependencyDeclaration()
        let result = declaration.DependsOnLazy<DynamicDependencyDeclarationTestsData.BaseModule>()
        do! check(Assert.That(Object.ReferenceEquals(result, declaration)).IsTrue())
    }

    [<Test>]
    member _.DependencyDeclaration_Throws_For_Non_Module_Type() = async {
        let declaration = DependencyDeclaration()
        let mutable message = null

        try
            declaration.DependsOn(typeof<string>) |> ignore
        with :? InvalidModuleTypeException as ex ->
            message <- ex.Message

        do! check(Assert.That(message <> null).IsTrue())
        do! check(Assert.That(message.Contains("is not a Module")).IsTrue())
    }

    [<Test>]
    member _.DependencyDeclaration_Required_Has_Correct_DependencyType() = async {
        let declaration = DependencyDeclaration()
        declaration.DependsOn<DynamicDependencyDeclarationTestsData.BaseModule>() |> ignore
        let deps = declaration.Dependencies
        do! check(IntEqualsAssertionExtensions.IsEqualTo(Assert.That(deps.Count), 1))
        do! check(Assert.That(deps.[0].Kind).IsEqualTo(DependencyType.Required))
        do! check(Assert.That(deps.[0].IsOptional).IsEqualTo(false))
    }

    [<Test>]
    member _.DependencyDeclaration_Optional_Has_Correct_DependencyType() = async {
        let declaration = DependencyDeclaration()
        declaration.DependsOnOptional<DynamicDependencyDeclarationTestsData.BaseModule>() |> ignore
        let deps = declaration.Dependencies
        do! check(IntEqualsAssertionExtensions.IsEqualTo(Assert.That(deps.Count), 1))
        do! check(Assert.That(deps.[0].Kind).IsEqualTo(DependencyType.Optional))
        do! check(Assert.That(deps.[0].IsOptional).IsEqualTo(true))
    }

    [<Test>]
    member _.DependencyDeclaration_Lazy_Has_Correct_DependencyType() = async {
        let declaration = DependencyDeclaration()
        declaration.DependsOnLazy<DynamicDependencyDeclarationTestsData.BaseModule>() |> ignore
        let deps = declaration.Dependencies

        do! check(IntEqualsAssertionExtensions.IsEqualTo(Assert.That(deps.Count), 1))
        do! check(Assert.That(deps.[0].Kind).IsEqualTo(DependencyType.Lazy))
        do! check(Assert.That(deps.[0].IsOptional).IsEqualTo(true))
    }

    [<Test>]
    member _.DependencyDeclaration_Conditional_Has_Correct_DependencyType() = async {
        let declaration = DependencyDeclaration()
        declaration.DependsOnIf<DynamicDependencyDeclarationTestsData.BaseModule>(true) |> ignore
        let deps = declaration.Dependencies

        do! check(IntEqualsAssertionExtensions.IsEqualTo(Assert.That(deps.Count), 1))
        do! check(Assert.That(deps.[0].Kind).IsEqualTo(DependencyType.Conditional))
        do! check(Assert.That(deps.[0].IsOptional).IsEqualTo(false))
    }

    [<Test>]
    member _.DependencyDeclaration_Conditional_False_Does_Not_Add_Dependency() = async {
        let declaration = DependencyDeclaration()
        declaration.DependsOnIf<DynamicDependencyDeclarationTestsData.BaseModule>(false) |> ignore
        let deps = declaration.Dependencies
        do! check(IntEqualsAssertionExtensions.IsEqualTo(Assert.That(deps.Count), 0))
    }
